### DISCUSSION TITLE
Cannot join, RX issues

### DISCUSSION THREAD

**Discussion by JF002** [Lora-net/LoRaMac-node#1308] [Q&A]
Hello!

I'm currently trying to port LoRaMac to the [Pine64 PineDio LoRa USB adapter](https://pine64.com/product/pinedio-usb-lora-adapter/). This is basically a board with a SX1262 and a CH341 USB to SPI converter.

So I implemented the RTC, SPI, GPIO,... porting layers and adapted them to the desktop PC environment. Now, I'm trying to run the LoRaMac "periodic-uplink-lpp" demo application. I'm using the TTN network.

I reached a point where my code sends the join-request message, TTN receives it and accepts it and forward the join-accept message to the gateway:
![image](https://user-images.githubusercontent.com/2261652/166160051-4b189b39-713d-4b9d-90b5-5879d7e38835.png)

However, my application have big troubles to receive those messages : most of the time, nothing is received and sometimes, it receives a message but fails at decoding it ([This check in soft-se.c fails](https://github.com/Lora-net/LoRaMac-node/blob/master/src/peripherals/soft-se/soft-se.c#L225)).
I checked with my SDR device : I can see the messages sent by my USB board and the one sent by the gateway. So, the join-accept message is sent by the GW but my USB board fails to receive it properly.

As most of RX errors seem to be caused by incorrect timing for RX1/RX2 windows, I changed the code so that the SX1262 would go into RX mode indefinitely as soon as TX Done is triggered. This way, I'm pretty sure the module is listening when the gateway sends the join-accept message, with no more luck, unfortunately.

So now, I'm looking for help : do you have any idea why RX does not work properly? Is there anything I can do to debug this issue?

Thanks !

**Comment by jamesl-dm** [comment]
Have you got the IQ polarity configured correctly?

Edit: sorry, stupid question. If you've ported LoRaMac-node, it will be setting the IQ polarity correctly already. I'd double check the encryption keys are right first. And check that the actual received bytes in the JoinAccept look sane. I think they're sent unencrypted, with just that CMAC check that you linked protecting them. So if the bytes other than the CMAC look sane, it will be an encryption issue.

**Comment by JF002** [comment]
Hi, and thank you so much for answering so quickly!

I confirm I'm using the driver for the sx1262 provided by LoRaMac-node, which seems to  configures the IQ polarity properly.

Regarding the encryption keys, the TTN console allows me to specify a device EUI, 'AppKey' and 'NwkKey'. I ensured that the device EUI, the AppKey and the NwkKey from the TTN console correspond to the one specified in [se-identity.h](https://github.com/Lora-net/LoRaMac-node/blob/master/src/peripherals/soft-se/se-identity.h#L74). The byte order seems to be correct, as TTN would report a "MIC mismatch" error when I revert them. I left all the other keys (J_S_INT_KEY, J_S_ENC_KEY, F_NWK_S_INT_KEY and many others) from this file to their default value.

I added a few debug logs to check the data sent and received. Here is the result :

```
###### ===================================== ######

Application name   : periodic-uplink-lpp
Application version: 1.3.0
GitHub base version: 4.6.0

###### ===================================== ######

GpioMcuSetInterrupt()
BoardGetUniqueId()
DevEui      : 70-B3-D5-7E-D0-04-F9-F0
JoinEui     : 00-00-00-00-00-00-00-00
Pin         : 00-00-00-00

SetPayload (23 B) :  0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0xf0 0xf9 0x04 0xd0 0x7e 0xd5 0xb3 0x70 0x01 0x00 0x06 0x42 0xe7 0x2b 

###### =========== MLME-Request ============ ######
######               MLME_JOIN               ######
###### ===================================== ######
IRQ : 1
IRQ : 4
IRQ : 16
OnRxDone IRQ (33 B, rssi = -34, snr = 11) :  0x00 0xca 0xf3 0xdf 0x7b 0xc8 0x69 0xdf 0xb1 0xfa 0x01 0xd2 0xce 0x03 0xe6 0x94 0x19 0x54 0xa9 0x41 0x73 0x29 0xdf 0x57 0xfe 0x90 0x19 0x30 0x8f 0xc8 0xbc 0x86 0x3e 

EepromMcuWriteBuffer()
EepromMcuWriteBuffer()
EepromMcuWriteBuffer()
EepromMcuWriteBuffer()
EepromMcuWriteBuffer()
EepromMcuWriteBuffer()

SetPayload (23 B) :  0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0xf0 0xf9 0x04 0xd0 0x7e 0xd5 0xb3 0x70 0x02 0x00 0xf8 0xed 0x98 0x37 

###### =========== MLME-Request ============ ######
######               MLME_JOIN               ######
###### ===================================== ######
IRQ : 1
IRQ : 4
IRQ : 16
IRQ : 2
OnRxDone IRQ (33 B, rssi = -35, snr = 9) :  0x20 0xc4 0x7c 0x4f 0x76 0xcf 0xd5 0x04 0x4d 0x3c 0x3b 0x23 0x09 0x05 0x2c 0x94 0x12 0x8d 0x52 0x93 0xce 0x43 0x86 0x1d 0x36 0x6a 0x99 0xbc 0xb7 0x2d 0xfe 0xec 0x3d 


###### =========== MLME-Confirm ============ ######
STATUS      : Join failed

```

- I added the log `SetPlayload()` in https://github.com/Lora-net/LoRaMac-node/blob/master/src/radio/sx126x/sx126x.c#L145
- `IRQ : ` in https://github.com/Lora-net/LoRaMac-node/blob/master/src/radio/sx126x/radio.c#L1258
- `OnRxDone IRQ` in https://github.com/Lora-net/LoRaMac-node/blob/master/src/radio/sx126x/radio.c#L1316

I've just noticed that the first time it attempts to join the network, the function `EepromMcuWriteBuffer()` is called and then the join procedure is started again without error. The 2nd time, the join request fails.

I don't know how to check that the data sent and received are sane. Could you link me to resources so I can learn to understand the content of the packets?

Thanks again for your help :+1:

**Reply by jamesl-dm** [reply]
> I don't know how to check that the data sent and received are sane. Could you link me to resources so I can learn to understand the content of the packets?

Sorry, seems like I remembered wrong. The join accept is encrypted - it's join request that's unencrypted. You can try to decrypt it with something like: https://github.com/jieter/python-lora (I haven't tried this one yet)

But that reminds me - decryption is done differently depending on the LoRaWAN version. And there have been a lot of mixups with LoRaWAN 1.1 encryption on various platforms. So it you've registered your device as 1.1, try registering it as 1.0.3 instead?

**Reply by JF002** [reply]
TTN allows me to choose the LoRaWAN version in the console:
![image](https://user-images.githubusercontent.com/2261652/166805979-6fac67dd-068a-4122-9a26-0bb1ff37bb72.png)

I can select LoRaWAN 1.0.0, 1.0.1, 1.0.2, 1.0.3, 1.0.4 and 1.1.0. When I select 1.0.0, I also need to choose  a Regional parameters version (1.0.0, 1.0.1, 1.0.2, 1.0.3, 1.1 rev A, 1.1 rev B).

Do I need to choose a LoRaWAN version according to the LoRacMAC version (which is 1.1.0, if I understand correctly)? Or will LoRaMAC adapt according to the lorawan version specific in TTN?

I did a few tests with lorawan 1.0.3 and 1.1.0 + RP 1.0.0 with similar results : 
 - When LoRaWAN version 1.0.3 is selected, JOIN fails here : https://github.com/Lora-net/LoRaMac-node/blob/master/src/peripherals/soft-se/soft-se.c#L361
 - And when LoRaWAN 1.1 is selected, it fails here : https://github.com/Lora-net/LoRaMac-node/blob/master/src/peripherals/soft-se/soft-se.c#L392
The error always comes from a failed aes cmac check.

I did notice that the data sent by my device is always the same: 

```
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0xf0 0xf9 0x04 0xd0 0x7e 0xd5 0xb3 0x70 0x01 0x00 0x06 0x42 0xe7 0x2b
```

However, the data received is always different. The processed and expected cmac value are also different, obviously. Here are a few examples:

```
#1
0x20 0x04 0x9e 0x8c 0x5e 0xdd 0x1e 0xf8 0xd4 0x6c 0xf3 0xa8 0x05 0x06 0x8a 0xf1 0x1b 0x9f 0x4e 0x08 0xd3 0x41 0xd5 0x9b 0x31 0x6a 0xea 0x7c 0x75 0xfe 0xc0 0x50 0x0f 
CMAC (expected / comp) : 1174086198 / 290745102

#2
0x20 0xa9 0x3d 0x45 0x4b 0x96 0xc5 0xf5 0x71 0x55 0x64 0xba 0x39 0xc1 0xbf 0x0d 0xc7 0xfc 0x81 0x9c 0x6a 0x65 0xd1 0x93 0xaa 0xb8 0x7c 0x70 0x86 0x06 0x04 0x55 0xa8 
CMAC (expected / comp) : 251006650 / 274756751

#3
0x20 0xa9 0x1e 0xb6 0xce 0xf9 0x36 0x75 0x84 0xa3 0xd4 0xb5 0x53 0xfe 0x84 0x87 0x6e 0x49 0xdd 0x17 0xc5 0xfc 0x49 0xf7 0xe2 0xd3 0x9c 0x28 0x82 0x9e 0xf0 0x5e 0x9c 
CMAC (expected / comp) : -1825622765 / -1211566570
```

Is there any issue on the RX side of my board? To check it, I created a very simple test program : 
```
  BoardInitMcu( );
  BoardInitPeriph( );

  RadioEvents_t event;
  event.RxDone = OnRxDone;
  Radio.Init(&event);

  Radio.SetChannel(868000000);
  Radio.SetRxConfig( MODEM_LORA, 2, 12, 1, 0, 8, 0xffff, false, 0, true, 0, 0, true, true );
  Radio.SetMaxPayloadLength( MODEM_LORA, 242 + LORAMAC_FRAME_PAYLOAD_OVERHEAD_SIZE );

  Radio.Rx(0xffffffff);

  while(1) {
    Radio.IrqProcess();
    usleep(1000);
  }
```

I could check with another device (a PyCom LOPY board as TX) that messages are correctly and reliably received by my board.

Thanks for helping! I hope we'll find what's wrong in my implementation!

**Reply by jamesl-dm** [reply]
Since it still fails the CMAC on 1.0, I guess we can be pretty sure it isn't an encryption compatibility issue related to the 1.1 encryption errata, or a problem with the nonces not incrementing. So the next thing to check is that the bytes are being received correctly. The downlink is the right size, and LoRa packets are protected by CRCs and forward error correction, so it's unlikely that the radio is receiving the wrong thing. But if your SPI bitrate is too high, or there is too much capacitance on the SPI bus (due to logic analyzer being connected for instance) you might occasionally read back the wrong bits.

I'd try setting the node up using ABP on LoRaWAN 1.0, so that you don't need to Join. Then you can send a downlink from the TTN console. The downlink appears twice in the console - first the unencrypted form, when the downlink is queued, then the encrypted form, when it gets sent to a gateway. You can check if the encrypted bytes are received. If the correct bytes are received, I'd start double checking the decryption code (does it use AES ENcryption to decrypt, are there any integer promotion bugs created by compiling 32 bit C on a 64 bit machine, is it getting the correct key from the keystore, endian mistakes in the nonces etc).

**Reply by JF002** [reply]
Regarding the SPI speed, it was running previously with (unknown to me) default values. I set the port speed to 100KHz but didn't notice any change.

I tried switching to ABP on LoRaWAN 1.0.3 and I queued a downlink message. Here is the TTN console:
![image](https://user-images.githubusercontent.com/2261652/167258541-b66b6f3d-ebbf-4ca0-b9de-d68938c30e70.png)

And here is the corresponding log from my application:
```
###### ===================================== ######

Application name   : periodic-uplink-lpp
Application version: 1.3.0
GitHub base version: 4.6.0

###### ===================================== ######

GpioMcuSetInterrupt()
BoardGetUniqueId()
EepromMcuReadBuffer()
EepromMcuReadBuffer()
EepromMcuReadBuffer()
EepromMcuReadBuffer()
EepromMcuReadBuffer()
EepromMcuReadBuffer()
EepromMcuReadBuffer()
EepromMcuReadBuffer()
DevEui      : 70-B3-D5-7E-D0-04-F9-F0
JoinEui     : 00-00-00-00-00-00-00-00
Pin         : 00-00-00-00

[1651933423,43206] DisplayMacMlmeRequestUpdate

###### =========== MLME-Request ============ ######
######               MLME_JOIN               ######
###### ===================================== ######
STATUS      : OK

###### =========== MLME-Confirm ============ ######
STATUS      : OK
###### ===========   JOINED     ============ ######

ABP

DevAddr     : 017A6112


EepromMcuWriteBuffer()
EepromMcuWriteBuffer()
EepromMcuWriteBuffer()
EepromMcuWriteBuffer()
EepromMcuWriteBuffer()
EepromMcuWriteBuffer()
SetPayload (20 B) :  
0x40 0x12 0x61 0x7a 0x01 0x80 0x01 0x00 0x02 0x40 0x51 0xcb 0x3f 0x33 0xd3 0xca 0x61 0xd2 0x55 0xd7 

###### =========== MCPS-Request ============ ######
######           MCPS_UNCONFIRMED            ######
###### ===================================== ######
STATUS      : OK
[1651933425,44585] IRQ
IRQ : 1
[1651933426,45621] OnRxWindow1TimerEvent
[1651933426,45792] IRQ
IRQ : 4
IRQ_PREAMBLE_DETECTED! 
[1651933426,46280] IRQ
IRQ : 16
IRQ_HEADER_VALID! 
[1651933427,46626] OnRxWindow2TimerEvent
Ignore RX2, RX1 is ongoing
[1651933428,47920] IRQ
IRQ : 2
RXDONE! 
OnRxDone IRQ (51 B, rssi = -17, snr = 10) :  
0x60 0x12 0x71 0x29 0x8b 0x90 0x00 0x00 0x30 0xf5 0x7d 0xf3 0xa0 0x84 0x8e 0x64 0x71 0x7b 0x29 0x2c 0x60 0xad 0xd6 0xb4 0x08 0x76 0xf0 0x0b 0x12 0xc6 0x87 0x01 0x53 0x2b 0xc6 0x0a 0x6c 0x59 0xfe 0xa2 0x90 0x80 0x4a 0x27 0x82 0xb8 0x3b 0xfd 0x0a 0xf0 0xb2 

###### =========== MCPS-Confirm ============ ######
STATUS      : OK

###### =====   UPLINK FRAME        1   ===== ######

CLASS       : A

TX PORT     : 2
TX DATA     : UNCONFIRMED
00 00 FF 01 02 42 68 

DATA RATE   : DR_0
U/L FREQ    : 868300000
TX POWER    : 0
CHANNEL MASK: 0007 
```

The packet containing `4051CB3F33D3CA` appears in the logs of my app : `SetPayload (20 B) :  
0x40 0x12 0x61 0x7a 0x01 0x80 0x01 0x00 0x02 0x40 0x51 0xcb 0x3f 0x33 0xd3 0xca 0x61 0xd2 0x55 0xd7 `.

However, the downlink message (unencrypted : 11223344 - encrypted 6A2C17CC) is not logged at all :/

By the way : the TX data is correct and correctly received by the TTN app.

Your comment about 32/64 bits issue is interesting! I'll check that out!

**Reply by jamesl-dm** [reply]
It looks like the downlink you received had the wrong DevAddr (8B297112 instead of 017A6112). Maybe it wasn't destined for your device?
![image](https://user-images.githubusercontent.com/28915246/167334022-6c688eff-438d-42dd-84e1-aa6c99a4276f.png)

I'd repeat the test, and see if the DevAddr is wrong a second time.
Here are three possibilities:
1.  The gateway is sending the right data, and the node is somehow misinterpreting it
2. The gateway is sending the right data, but the node is turning its receiver on too late, and you're receiving random packets meant for other devices instead
3. The network is sending the data to the gateway, and it's arriving too late and being dropped, and then you're receiving random packets meant for other devices instead. You can check this using the gateway logs. But it sounds like you already checked with an SDR. Were the received packets on the SDR at the correct timing?

**Reply by JF002** [reply]
Hi, and sorry for the delay of my answer!

So, I ran it again:
TX1 : `0x80 0x12 0x61 0x7a 0x01 0x80 0x01 0x00 0x02 0x40 0x51 0xcb 0x3f 0x33 0xd3 0xca 0x14 0x6b 0xe8 0xeb `
RX1 : `0x60 0x12 0x71 0x1a 0x0b 0x90 0x00 0x00 0x30 0xf5 0x75 0xf3 0xa0 0x84 0x8e 0x64 0x71 0xfb 0xe9 0x2c 0x60 0x2d 0xd6 0xb4 0x08 0x76 0xf0 0x1b 0x12 0xc6 0x87 0x01 0x53 0x22 0xc6 0x5b 0x6d 0x59 0x6e 0xc1 0xd2 0xf0 0x4a 0x27 0x81 0xaa 0x0b 0x7f 0x28 0xae 0x05`
TX2 : `0x80 0x12 0x61 0x7a 0x01 0x80 0x02 0x00 0x02 0x06 0x1f 0xe2 0x85 0x07 0x52 0x01 0xb6 0xf3 0x1d 0xdb `
RX2 : `0x60 0x12 0x61 0x1f 0x8b 0xb0 0x01 0x00 0x56 0xbd 0xf6 0x6c 0xcf 0xd0 0xb3 0x7b 0xe7 0x0b 0x72 0x54 0x7c 0x8d 0xa0 0x57 0xed 0x88 0xd9 0x4d 0x31 0x7d 0xb3 0x93 0x51 0x6b 0xd3 0x9a 0x18 0x1c 0x6b 0x8e 0x8f 0xd2 0xaf 0x95 0x8b 0xc3 0xce 0x18 0x23 0x30 0x89`
TX3 : `0x80 0x12 0x61 0x7a 0x01 0x80 0x03 0x00 0x02 0xa1 0x4b 0x54 0x27 0xf7 0x39 0x5d 0xb9 0xc6 0x13 0xd9`
RX3 : `0x60 0x12 0x71 0x29 0x8b 0xb0 0x02 0x00 0x50 0xd3 0x53 0x38 0x39 0x3a 0x7a 0x39 0x9d 0x06 0x7a 0xee 0xba 0x06 0xd2 0xf5 0x1f 0xf8 0x70 0xd4 0xb3 0xe1 0x93 0xcd 0xc4 0xad 0x33 0x0a 0x69 0xa5 0x9e 0xcb 0x9e 0xe3 0xd7 0x53 0x4e 0xf5 0x72 0x99 0xb8 0xaa 0xf9 `

I found the corresponding logs in the gateway:
```
JSON up: {"rxpk":[{"jver":1,"tmst":804120516,"time":"2022-05-16T19:48:55.120515Z","tmms":1336765753120,"chan":2,"rfch":1,"freq":868.500000,"mid": 0,"stat":1,"modu":"LORA","datr":"SF12BW125","codr":"4/5","rssis":-66,"lsnr":6.5,"foff":-11210,"rssi":-66,"size":20,"data":"gBJhegGAAQACQFHLPzPTyhRr6Os="}]}

JSON down: {"txpk":{"imme":false,"tmst":805120516,"freq":868.5,"rfch":0,"powe":14,"modu":"LORA","datr":"SF12BW125","codr":"4/5","ipol":true,"size":51,"ncrc":true,"data":"YBJhegGwAAAA/XHzsOfCZHjr6ibgbcaHgrT4G3EJjwFTEgDKbFldB9D6SnSEnpt/Go5N"}}

JSON up: {"rxpk":[{"jver":1,"tmst":812329461,"time":"2022-05-16T19:49:11.329461Z","tmms":1336765769329,"chan":1,"rfch":1,"freq":868.300000,"mid": 0,"stat":1,"modu":"LORA","datr":"SF12BW125","codr":"4/5","rssis":-64,"lsnr":9.2,"foff":-11138,"rssi":-64,"size":20,"data":"gBJhegGAAgACBh/ihQdSAbbzHds="}]}

JSON down: {"txpk":{"imme":false,"tmst":813329461,"freq":868.3,"rfch":0,"powe":14,"modu":"LORA","datr":"SF12BW125","codr":"4/5","ipol":true,"size":51,"ncrc":true,"data":"YBJhegGwAQAAMfZoz7K/++Ybcd5wjaAOqArZTZixO5NRXnYeGBxaQqfar7VnU8pYY6wL"}}

JSON up: {"rxpk":[{"jver":1,"tmst":820539002,"time":"2022-05-16T19:49:27.539001Z","tmms":1336765785539,"chan":1,"rfch":1,"freq":868.300000,"mid": 0,"stat":1,"modu":"LORA","datr":"SF12BW125","codr":"4/5","rssis":-66,"lsnr":8.0,"foff":-11081,"rssi":-65,"size":20,"data":"gBJhegGAAwACoUtUJ/c5XbnGE9k="}]}

JSON down: {"txpk":{"imme":false,"tmst":821539002,"freq":868.3,"rfch":0,"powe":14,"modu":"LORA","datr":"SF12BW125","codr":"4/5","ipol":true,"size":51,"ncrc":true,"data":"YBJhegGwAgAA/1k6KXlSMZkWGu6nBtKWWfAw1IAtnc3EjPTIaaWvDz7j1yDyvXKpGWB7"}}
```

The data sent by the device and the data received (and BASE64 decoded) by the gateway (JSON UP) are equal. Ex:
```
gBJhegGAAQACQFHLPzPTyhRr6Os= -base64-> 8012617a01800100024051cb3f33d3ca146be8eb == TX1
```

But the data received by the gateway from TTN (JSON DOWN) and the data received by the device are NOT equal:
```
YBJhegGwAAAA/XHzsOfCZHjr6ibgbcaHgrT4G3EJjwFTEgDKbFldB9D6SnSEnpt/Go5N -base64-> 6012617a01b0000000fd71f3b0e7c26478ebea26e06dc68782b4f81b71098f01531200ca6c595d07d0fa4a74849e9b7f1a8e4d != RX1
```

So uh... it looks like my device does not receive the LoRaWAN packets correctly? This is confusing as my simple test program (raw LoRa RX using the sx1262 driver from loramac) [from this comment](https://github.com/Lora-net/LoRaMac-node/discussions/1308#discussioncomment-2689425) is still working properly and reliably.

**Reply by jamesl-dm** [reply]
Yeah that looks very suspicous.

```
received  : 60 12711a0b 90 0000 30 f575f3a0848e6471fbe92c602dd6b40876f01b12c687015322c65b6d596ec1d2f04a2781aa0b 7f28ae05
correct   : 60 12617a01 b0 0000 00 fd71f3b0e7c26478ebea26e06dc68782b4f81b71098f01531200ca6c595d07d0fa4a74849e9b 7f1a8e4d
difference: 00 0010600a 20 0000 30 08040010634c000910030a804010338ac2080063cf08000030c691010033c6020a0053053490 00322048
```

If these were different packets, the last two fields would be random. But they differ in less than a quarter of their bits - so strongly correlated.  Since your test program was getting the right values, let's assume the SPI communications are OK.

The next thing I would suspect is the crystal. Is it a crystal, not a TCXO? You could try running the test under a fast fan, to keep the temperature as constant as possible, and you could try using the lowest transmit power to reduce heat generation.

In the test against the network, you transmit and then receive. While the transmitter is running, the SX1262 is going to heat up a bit, and the crystal next to it will heat up a bit and shift frequency. Then when it starts receiving, the crystal will be cooling down. It generates the most heat, and is the most sensitive to heat, at SF12, which you're using in the network test. It's well known that you can run into issues using a crystal instead of a TCXO at SF12, unless you're careful with the thermal design. But most devices use a TCXO. If your adapter IS using a TCXO, check that the TCXO voltage is configured correctly.

**Reply by JF002** [reply]
I don't really know if the board uses a crystal of a TCXO. Can we deduce that from the [schematics](https://wiki.pine64.org/images/4/4d/USB_LoRa_adapter-PINE64_RFM90_SCH_2021_04_26.pdf)?
I did a few tests with a fan blowing on the board, with not much success, unfortunately :/
Is it possible to configure LoRaMAC to use another SF than SF12 by default? 

I'll do a few SPI check (read/write in the payload buffer) to ensure that SPI communication is reliable.

**Reply by jamesl-dm** [reply]
Yeah it looks like it's just a crystal (to double check, see if component MFB1 is populated - no MFB1 = no TCXO). No sure why it's marked 32.728 kHz on the schematics - should be 32 MHz. Try backing the power level off to reduce heat buildup. Also, check the center frequency is accurate, since they might have got the crystal capacitors wrong. You can do that by calibrating your SDR's ppm against a known source (ie. using Kalibrate or LTE-Cell-Scanner), then producing some continuous wave output from your pine. You should be able to see the initial offset, and some drift from heat buildup.

**Reply by jamesl-dm** [reply]
I guess it's worth putting it in ABP and setting the datarate up to SF7 as well? If you can get downlinks at SF7, but not at SF12, you can be pretty confident its a crystal problem.

**Reply by JF002** [reply]
I have a very good news! I asked my friend @lupyuen (who have already worked with LoRaMAC on other devices and who owns the same board than me) to help me diagnose this issue, and we've eventually managed to reliably connect to LoRaWAN networks (local Chirpstack instance and public TTN).
There are apparently 2 issues : the timing for the RX window that needed to be customized and the SPI driver that seems to work much better when reading the data by chunks instead of reading byte-per-byte.

Now I feel embarrassed as you've already mention those as possible cause of my issues and I couldn't find any problem around them... Sorry about that!

You'll find the results of our experiments [here](https://codeberg.org/JF002/loramac-node/pulls/1). The code is quite messy and needs to be cleaned and improved, but I'm already very happy we managed to connect to LoRaWAN networks using this USB adapter!

Thank you very much for you help, time and patience!

**Comment by MedadRufus** [comment]
If you are concerned about timing accuracy for the RX window, set the `LmHandlerSetSystemMaxRxError(20);` to something larger, like 500 milliseconds, as I have done in this pull request:
https://github.com/ImperialSpaceSociety/LoRaMac-node/commit/cf4cf9ac3a068c7cd1693ecc47f2d6373a37b287#diff-169f8b670ece097ea0c2244bdcbc2c51a64fbb68473609a83c58f39516d37141R262

This is the safest way to handle timing inaccuracies without potentially breaking code changes.

**Reply by JF002** [reply]
Thanks, good idea. I'll give it a shot!

**Comment by MedadRufus** [comment]
Also, have you gotten loramac node and your new credentials working on one of the project's supported boards? It will allow you to narrow down your problems.

**Reply by JF002** [reply]
I do not have any device supported by LoRaMAC on hands. However, I've already managed to connect a PyCom LOPY board and a TTGO TTNMapper device to my TTN account.
